### PR TITLE
intake

### PR DESCRIPTION
**Pull request by PotmanNob**
w intake trust trust

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Intake/Constants.java
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/Constants.java:1]: Rename to IntakeConstants

File: src/main/java/frc/robot/subsystems/Intake/Constants.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    public final static double IDLE_PIVOT = 0;
+
+    public final static int NUM_PIVOT_MOTORS = 1;
+    public final static double PIVOT_GEARING = 5;
+    public final static double PIVOT_MOI = 0.31654227;
+    public final static double PIVOT_ARM_LENGTH = 26.199558;
+    public final static double MIN_PIVOT_ANGLE = 0;
+    public final static double MAX_PIVOT_ANGLE = 70;
+    public final static double STARTING_PIVOT_ANGLE = 70;
+    public final static PIDConstants PIVOT_PID_CONSTANTS = new PIDConstants(0, 0, 0);
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/Constants.java:23]: guess and check PID with sim

File: src/main/java/frc/robot/subsystems/Intake/IntakeIOSim.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.simulation.DCMotorSim;
+import edu.wpi.first.wpilibj.simulation.SingleJointedArmSim;
+import edu.wpi.first.math.system.plant.LinearSystemId;
+
+public class IntakeIOSim implements IntakeIO {
+    SingleJointedArmSim pivotSim;
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/IntakeIOSim.java:12]: make all these private

**Review by GreenTomato5 — CHANGES_REQUESTED** [review]
mostly fire

File: src/main/java/frc/robot/subsystems/Intake/IntakeIOReal.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        pivotMotor = new TalonFX(IntakeConstants.Real.PIVOT_MOTOR_CANID);
+        pivotController = new PIDController(IntakeConstants.Real.PIVOT_PID_CONSTANTS.kP,
+        IntakeConstants.Real.PIVOT_PID_CONSTANTS.kI, IntakeConstants.Real.PIVOT_PID_CONSTANTS.kD);
+        wheelSpeedController = new PIDController(IntakeConstants.Real.WHEEL_PID_CONSTANTS.kP,
+        IntakeConstants.Real.WHEEL_PID_CONSTANTS.kI, IntakeConstants.Real.WHEEL_PID_CONSTANTS.kD);
+    }
+
+    @Override
+    public void updateInputs(IntakeIOInputs inputs) {
+        inputs.pivotPosition = Units.rotationsToDegrees(pivotMotor.getPosition().getValueAsDouble());
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/IntakeIOReal.java:27]: Specifically for the IO real, use smart dashboard to log the PID controller if you're in testing mode (Should be renamed to tuning if its not already)

File: src/main/java/frc/robot/subsystems/Intake/IntakeIOReal.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    private TalonFX wheelMotor;
+    private TalonFX pivotMotor;
+    private PIDController pivotController;
+    private PIDController wheelSpeedController;
+    private double pivotPositionSetpoint;
+    private double wheelSpeedSetpoint;
+
+    public IntakeIOReal() {
+        wheelMotor = new TalonFX(IntakeConstants.Real.WHEEL_MOTOR_CANID);
+        pivotMotor = new TalonFX(IntakeConstants.Real.PIVOT_MOTOR_CANID);
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/IntakeIOReal.java:18]: you have to zero the encoder for the pivot should be pivotMotor.setPosition(0)

File: src/main/java/frc/robot/subsystems/Intake/IntakeIOSim.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import edu.wpi.first.wpilibj.simulation.DCMotorSim;
+import edu.wpi.first.wpilibj.simulation.SingleJointedArmSim;
+import edu.wpi.first.math.system.plant.LinearSystemId;
+
+public class IntakeIOSim implements IntakeIO {
+    private SingleJointedArmSim pivotSim;
+    private DCMotorSim wheelMotorSim;
+    private PIDController pivotController;
+    private PIDController wheelSpeedController;
+
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/IntakeIOSim.java:18]: If you don't want to add ctre sim with the wpilib physics sim open an issue about it so sum1 else can do it once this is merged

File: src/main/java/frc/robot/subsystems/Intake/IntakeStates.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+import frc.robot.pioneersLib.subsystem.SubsystemStates;
+
+public enum IntakeStates implements SubsystemStates {
+    INTAKING(IntakeConstants.INTAKING_SPEED.magnitude(), IntakeConstants.INTAKING_PIVOT.magnitude(), "INTAKING"),
+    OUTTAKING(IntakeConstants.OUTTAKING_SPEED.magnitude(), IntakeConstants.OUTTAKING_PIVOT.magnitude(), "OUTTAKING"),
+    PASSING(IntakeConstants.PASSING_SPEED.magnitude(), IntakeConstants.PASSING_PIVOT.magnitude(), "PASSING"),
+    IDLE(IntakeConstants.IDLE_SPEED.magnitude(), IntakeConstants.IDLE_PIVOT.magnitude(), "IDLE");
+
+    IntakeStates(Double wheelSpeed, double pivotSetpoint, String stateString) {
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/IntakeStates.java:11]: Ok buddy put state string first

File: src/main/java/frc/robot/subsystems/Manager/ManagerStates.java [hunk 7]
```diff
@@ -1,18 +1,26 @@
...
 
     private String stateString;
+    private IntakeStates intakeState;
 
     @Override
     public String getStateString() {
         return stateString;
     }
+
+    public IntakeStates getIntakeState() {
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Manager/ManagerStates.java:23]: cant this be protected?

**Review by GreenTomato5 — APPROVED** [review]
Best code I've ever seen

**Review by GreenTomato5 — CHANGES_REQUESTED** [review]
I meant to request c hanges (see the comments)

File: src/main/java/frc/robot/subsystems/Intake/IntakeStates.java (hunk 6 above)
Comment (PotmanNob) [reply src/main/java/frc/robot/subsystems/Intake/IntakeStates.java:11]: buddy is a hater

File: src/main/java/frc/robot/subsystems/Intake/Intake.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    
+    private Intake(IntakeIO io) {
+        super("Intake", IntakeStates.IDLE);
+        this.io = io;
+        inputs = new IntakeIOInputsAutoLogged();
+    }
+
+    @Override
+    protected void runState() {
+        Logger.recordOutput("Intake State", getState());
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/Intake.java:22]: this logs output to the "recorded outputs" dir in ascope, you should do SUBSYSTEM_NAME/ + "Intake State" and store subsystem name in constants. Same for all other logs

File: src/main/java/frc/robot/subsystems/Intake/IntakeIOReal.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        wheelMotor.setPosition(0);
+        pivotMotor.setPosition(0);
+        pivotController = new PIDController(IntakeConstants.Real.PIVOT_PID_CONSTANTS.kP,
+        IntakeConstants.Real.PIVOT_PID_CONSTANTS.kI, IntakeConstants.Real.PIVOT_PID_CONSTANTS.kD);
+        wheelSpeedController = new PIDController(IntakeConstants.Real.WHEEL_PID_CONSTANTS.kP,
+        IntakeConstants.Real.WHEEL_PID_CONSTANTS.kI, IntakeConstants.Real.WHEEL_PID_CONSTANTS.kD);
+
+        if (GlobalConstants.ROBOT_MODE == RobotMode.TESTING) {
+            SendableRegistry.setName(pivotController, "Intake", "Pivot PIDController");
```
Comment (GreenTomato5) [inline src/main/java/frc/robot/subsystems/Intake/IntakeIOReal.java:30]: Check the thing I said in call?

**Review by GreenTomato5 — CHANGES_REQUESTED** [review]
Almost!

**Review by GreenTomato5 — APPROVED** [review]
Work on your commit messages

**Pull request merged**